src/python/pants/backend/code_quality_tool/lib.py

```diff
@@ -0,0 +1,418 @@
+
+    extra_env: dict[str, str] = dict(run_request.extra_env or {})
+    extra_path = extra_env.pop("PATH", None)
+
+    extra_sandbox_contents = []
+
+    extra_sandbox_contents.append(
+        ExtraSandboxContents(
+            EMPTY_DIGEST,
+            extra_path,
+            run_request.immutable_input_digests or FrozenDict(),
+            run_request.append_only_caches or FrozenDict(),
+            run_request.extra_env or FrozenDict(),
+        )
+    )
+
+    if runnable_dependencies:
+        extra_sandbox_contents.append(
+            ExtraSandboxContents(
+                EMPTY_DIGEST,
+                f"{{chroot}}/{runnable_dependencies.path_component}",
+                runnable_dependencies.immutable_input_digests,
+                runnable_dependencies.append_only_caches,
+                runnable_dependencies.extra_env,
+            )
+        )
+
+    merged_extras = await Get(
+        ExtraSandboxContents, MergeExtraSandboxContents(tuple(extra_sandbox_contents))
+    )
+    extra_env = dict(merged_extras.extra_env)
+    if merged_extras.path:
+        extra_env["PATH"] = merged_extras.path
+
+    digest = await Get(Digest, MergeDigests((dependencies_digest, run_request.digest)))
+
+    return CodeQualityToolBatchRunner(
+        digest=digest,
+        args=run_request.args + tuple(cqt.args),
+        extra_env=FrozenDict(extra_env),
+        append_only_caches=merged_extras.append_only_caches,
+        immutable_input_digests=merged_extras.immutable_input_digests,
+    )
+
+
+@dataclass
+class CodeQualityToolRuleBuilder:
+    goal: str
+    target: str
+    name: str
+    scope: str
+
+    def _build_lint_rules(self) -> Iterable[Rule]:
+        class CodeQualityToolInstance(Subsystem):
+            options_scope = self.scope
+            name = self.name
+            help = f"{self.goal.capitalize()} with {self.name}. Tool defined in {self.target}"
+
+            skip = SkipOption("lint")
+
+        class CodeQualityProcessingRequest(LintFilesRequest):
+            tool_subsystem = CodeQualityToolInstance
+
+        @rule(canonical_name_suffix=self.scope)
+        async def partition_inputs(
+            request: CodeQualityProcessingRequest.PartitionRequest,
+            subsystem: CodeQualityToolInstance,
+        ) -> Partitions:
+            if subsystem.skip:
+                return Partitions()
+
+            cqt = await Get(CodeQualityTool, CodeQualityToolAddressString(address=self.target))
+
+            matching_filepaths = FilespecMatcher(
+                includes=cqt.file_glob_include,
+                excludes=cqt.file_glob_exclude,
+            ).matches(request.files)
+
+            return Partitions.single_partition(sorted(matching_filepaths))
+
+        @rule(canonical_name_suffix=self.scope)
+        async def run_code_quality(request: CodeQualityProcessingRequest.Batch) -> LintResult:
+            sources_snapshot = await Get(Snapshot, PathGlobs(request.elements))
```

Comment:
I don't recall the interaction of the file watcher and this sort of globbing. If one has something like `pants --loop lint ::`  running, and then makes a change to a relevant file, does that change correctly retrigger this rule to run?

Reply:
I would think/guess so. Not 100% sure, ngl. 😂

Reply:
Ah cool, thanks for checking. Maybe this is #18888 for the lint goal too.